### PR TITLE
Add undo/redo functionality

### PR DESCRIPTION
**Pull request by suricactus**
This PR adds undo/redo functionality on adding/editing/deleting layer features.

The undo/redo buttons can be found at the top of the the main menu in top right.

All feature modifications are stored in undo or redo stacks with the most recent modification on top. The undo and redo stacks are stored in memory, so they will be lost on restarting the app.

Whenever a feature is modified, a snapshot of the feature's old and new state is stored in the undo stack of the device memory and  the redo stack is emptied.

When undo menu icon is pressed, the most recent entry from the undo stack is popped and the feature gets it's previous state. A new redo stack entry is added.

When redo menu icon is pressed, the most recent  entry from the redo stack is popped and the feature gets it's newer state. A new undo stack entry is added.

Recording 2023-12-12 at 23 04 21

~TODO describe this PR~ Done
~TODO fix the icons on dark theme~ Done
~TODO name icons as 24px and white~ Done

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 8833d9f354f8736c8efb0fd912f5d3e9e15cca6f: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4849-f27ff7bc5128bf6f2a43d23c956b0ba41c9b9879-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4849-f27ff7bc5128bf6f2a43d23c956b0ba41c9b9879-arm-neon-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4849-f27ff7bc5128bf6f2a43d23c956b0ba41c9b9879-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4849-f27ff7bc5128bf6f2a43d23c956b0ba41c9b9879-x86-android.apk)
- [**all access arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4849-f27ff7bc5128bf6f2a43d23c956b0ba41c9b9879-arm64-android.apk)
- [all access arm-neon-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4849-f27ff7bc5128bf6f2a43d23c956b0ba41c9b9879-arm-neon-android.apk)
- [all access x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4849-f27ff7bc5128bf6f2a43d23c956b0ba41c9b9879-x64-android.apk)
- [all access x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4849-f27ff7bc5128bf6f2a43d23c956b0ba41c9b9879-x86-android.apk)
</details>

File: src/core/featurehistory.cpp [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+  for ( const QgsFeatureId fid : changedAttributesFids )
+    changedFids.insert( fid );
+
+  // NOTE we read the features from the dataProvider directly as we want to access the old values.
+  // If we use the layer, we get the values from the edit buffer.
+  QgsFeatureIterator featuresIt = vl->dataProvider()->getFeatures( QgsFeatureRequest( changedFids ) );
+  QMap<QgsFeatureId, QgsFeature> modifiedFeatures;
+  QgsFeature f;
+
+  // ? is it possible to use the iterator in a less ugly way? something like normal `for ( QgsFeature &f : it ) {}`
```
Comment (nirvn) [inline src/core/featurehistory.cpp:103]: I'm used to this while form; beauty is in the eye of the beholder, I see no uglyness :wink:

File: src/core/featurehistory.cpp [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+  }
+
+  const QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( sender() );
+
+  if ( !vl )
+  {
+    return;
+  }
+
+  qInfo() << "FeatureHistory::onCommittedFeaturesAdded: adding create committed features";
```
Comment (nirvn) [inline src/core/featurehistory.cpp:128]: I'm not sure we want this here, it'll end up spamming the sentry logs, maybe make these qDebug so it's not showing in release builds.

File: src/core/featurehistory.cpp [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    if ( !vl->commitChanges() )
+    {
+      QgsMessageLog::logMessage( tr( "Failed to commit undo feature modification in layer \"%1\"" ).arg( vl->name() ) );
+
+      if ( !vl->rollBack() )
+      {
+        QgsMessageLog::logMessage( tr( "Failed to rollback undo featurue modifications in layer \"%1\"" ).arg( vl->name() ) );
+      }
+
+      // OGR error creating feature -4: failed to execute insert : UNIQUE constraint failed: area.fid
```
Comment (nirvn) [inline src/core/featurehistory.cpp:323]: What's that?

File: src/core/featurehistory.h [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+#include <qgsproject.h>
+
+
+typedef QPair<QgsFeature, QgsFeature> OldNewFeaturePair;
+
+class FeatureHistory : public QObject
+{
+    Q_OBJECT
+
+    Q_PROPERTY( bool isUndoEnabled READ isUndoEnabled NOTIFY isUndoEnabledChanged )
```
Comment (nirvn) [inline src/core/featurehistory.h:15]: I think I'd rename this to is{Undo, Redo}Available

File: src/core/featurehistory.h [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+#define FEATUREHISTORY_H
+
+#include <QObject>
+#include <QTimer>
+#include <qgsproject.h>
+
+
+typedef QPair<QgsFeature, QgsFeature> OldNewFeaturePair;
+
+class FeatureHistory : public QObject
```
Comment (nirvn) [inline src/core/featurehistory.h:11]: @suricactus , big thanks for the documentation here, you're a role model to all of us :wink:

File: src/core/featurehistory.h [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    //! Undo history records
+    QList<QMap<QString, FeatureModifications>> mUndoHistory;
+
+    //! Redo history records
+    QList<QMap<QString, FeatureModifications>> mRedoHistory;
+
+    //! Layer ids being observed for changes. Should reset when the project is changed. Used to prevent double event listeners.
+    QSet<QString> mObservedLayerIds;
+
+  signals:
```
Comment (nirvn) [inline src/core/featurehistory.h:93]: Throughout QGIS and our code, we put signals and private slots _above private members.

File: src/qml/qgismobileapp.qml [hunk 7]
```diff
@@ -1249,6 +1249,7 @@ ApplicationWindow {
     spacing: 4
 
     QfToolButtonDrawer {
+      id: mainToolbarDrawer
```
Comment (nirvn) [inline src/qml/qgismobileapp.qml:1252]: Is that change necessary?

File: src/qml/qgismobileapp.qml [hunk 8]
```diff
@@ -2116,6 +2117,48 @@ ApplicationWindow {
         return result + padding * 2;
     }
 
+    Row {
+      bottomPadding: parent.topMargin
+
+      QfToolButton {
+        icon.source: Theme.getThemeVectorIcon( "ic_undo" )
```
Comment (nirvn) [inline src/qml/qgismobileapp.qml:2124]: The icons here should be named following the following standard: ic_{blah}_{black,white}_{24,48,etc.}. In your case it'd be ic_{undo,redo}_black_24dp

**Review by nirvn — COMMENTED** [review]
Nice, simple, functional.

File: src/core/featurehistory.cpp (hunk 1 above)
Comment (suricactus) [reply src/core/featurehistory.cpp:103]: it's a matter of taste for sure.

File: src/core/featurehistory.cpp (hunk 2 above)
Comment (suricactus) [reply src/core/featurehistory.cpp:128]: It's copy-paste from the layer observer. Changed to qDebug anyways here. We should not change layer observer for now I would say.

File: src/core/featurehistory.cpp (hunk 3 above)
Comment (suricactus) [reply src/core/featurehistory.cpp:323]: You just found one of the error messages that I had to debug :) . Deleted the line.

File: src/core/featurehistory.h (hunk 4 above)
Comment (suricactus) [reply src/core/featurehistory.h:15]: Renamed.

File: src/qml/qgismobileapp.qml (hunk 7 above)
Comment (suricactus) [reply src/qml/qgismobileapp.qml:1252]: Not required, tested some things. IMO it does not hurt, but will remove it upon your request.

File: src/core/featurehistory.h (hunk 5 above)
Comment (suricactus) [reply src/core/featurehistory.h:11]: :) just making my future life easier.

**Comment by nirvn** [comment]
One thing I'm just realizing we still need to address: how to handle tracking sessions. We need a mechanism to toggle off the observation of a given layer if it's currently tracking position. Otherwise we'll end up with an undo redo stack that's full of tracking feature updates that aren't really user actions.

**Comment by suricactus** [comment]
Good remark! Will address these days.

**Comment by suricactus** [comment]
Most probably the tracking issue is fixed now, needs to be tested and verified before merging.

File: src/core/featurehistory.cpp [hunk 9]
```diff
@@ -72,6 +74,12 @@ void FeatureHistory::onBeforeCommitChanges()
     return;
   }
 
+  Tracker *layerTracker = mTrackingModel->trackerForLayer( vl );
```
Comment (nirvn) [inline src/core/featurehistory.cpp:77]: I think we should approach this slightly differently and connect to the layerInTrackingChanged( qgsvectorlayer *layer, bool tracking ) signal. When we get a layer tracking = true, we disconnect the layer signals (beforeCommitChanges,afterCommitChanges,committedFeaturesAdded,committedFeaturesRemoved, etc) and when the tracking is turned off, we reconnect.

Otherwise we'll still end up responding to quite a lot of signals. When debugging sensor stuff, where tracking can occur multiple times per second, I became quite aware of the cost of signals :)

Comment (suricactus) [reply src/core/featurehistory.cpp]: I considered this as premature optimization during development and also assumed C++ signals are quite cheap. Seems I have been wrong, so I accommodated the requested changes.

Comment (suricactus) [reply src/core/featurehistory.cpp]: Tested with TCP server tracking, seems it works as expected. 
- tracking enabled - no undo/redo, even for manually digitezed features
- tracking disabled - all good

File: src/core/featurehistory.cpp [hunk 10]
```diff
@@ -214,6 +198,24 @@ void FeatureHistory::onLayersAdded( const QList<QgsMapLayer *> &layers )
   addLayerListeners();
 }
 
+void FeatureHistory::onLayerInTrackingChanged( QgsVectorLayer *vl, bool isTracking )
+{
+  qDebug() << "START TRACKING" << vl->id() << isTracking;
```
Comment (nirvn) [inline src/core/featurehistory.cpp:203]: Let's remove this.

**Review by nirvn — APPROVED** [review]
@suricactus , just one comment to address, then this looks to me like ready to merge, fantastic work!

File: src/core/featurehistory.cpp (hunk 10 above)
Comment (suricactus) [reply src/core/featurehistory.cpp]: Thanks for noticing.
